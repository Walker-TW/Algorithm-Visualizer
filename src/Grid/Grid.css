.Column {
  display: inline-block;
}
.Grid {
  margin: 40px 50% 0 0;
  width: 100%;
}
