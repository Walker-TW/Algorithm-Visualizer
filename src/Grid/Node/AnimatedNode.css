.Node {
  width: 20px;
  height: 20px;
  border: 1px solid black;
  background-color: grey;
  transition: transform 0.2s;
}

.visited {
  animation-name: visitedAnimation;
  animation-duration: 0.7s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes visitedAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background: radial-gradient(circle, #739da5, #020024, #090979);
  }
}

.start {
  background-color: rgb(206, 47, 255);
  animation-name: startAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes startAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: rgb(206, 47, 255);
  }
}

.finish {
  background-color: yellow;
  animation-name: finishAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes finishAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: yellow;
  }
}

.path {
  background-color: chartreuse;
  animation-name: pathAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes pathAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(186, 24, 178);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(247, 191, 70);
  }

  100% {
    transform: scale(1);
    background-color: chartreuse;
  }
}

.fence {
  animation-name: fenceAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}

@keyframes fenceAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: rgb(0, 0, 0);
  }
}
