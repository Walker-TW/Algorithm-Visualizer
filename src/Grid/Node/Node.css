.Node {
  width: 20px;
  height: 20px;
  border: 1px solid black;
  background-color: lightcyan;
}

.visited {
  background-color: hotpink;
}

.start {
  background-color: rgb(206, 47, 255);
}

.finish {
  background-color: yellow;
}

.path {
  background-color: chartreuse;
}
