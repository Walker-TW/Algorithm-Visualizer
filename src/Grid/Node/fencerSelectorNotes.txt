Need to do two things
- Create button that will switch the mouse clicks to walls 

- OR make it so that mouseenter etc events set start and finish nodes
- - And the double click on a node will change it to a start/finish node
- - Thing to make sure of here is that the wall selector is 
automatically set to false when this happens

- - MouseEnter would set a boolean that the mouse is clicked 
so we can use mouseenter to get inside and
toggle it 
- - that way when you lift up on the click it mouseisclicked will 
default to true and not be able to
make walls anymore (allowing you to switch between them both easily)


Pseudo

- Button Toggle
- - fenceToggle DEFAULT false X
- - onClick == function (which turns this true in the grid state) X

- On mousedown X
- - set fenceSelector = true
- - WHEN fenceSelector.true & button toggle has happened
- - THEN set fence = true

- On mouseenter X
- - if fenceSelector = true & button toggle 
- - set fence = true

- On mouseup X
- - set fenceSelector = false