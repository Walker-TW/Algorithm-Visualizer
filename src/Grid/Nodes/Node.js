import React from 'react';

const Node = () => {
  const getStyle = {
    width: '20px',
    height: '20px',
    border: '1px solid black',
  };

  return <div style={getStyle} />;
};

export default Node;
