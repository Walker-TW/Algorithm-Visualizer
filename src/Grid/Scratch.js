function dijkstra(startNode, endNode, grid) {
  const potentialRoute = [];
  const unvisitedNodes = grid;

  const neareastNeighbour = unvisitedNodes.shift();

  while (neareastNeighbour !== Infinity) {}
}
