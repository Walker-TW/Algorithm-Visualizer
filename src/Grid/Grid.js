import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";

import Node from "./Node/Node";
import Header from "./Header/Header";
import { Alert } from "react-bootstrap";
import Info from "./Info/Info";
import { dijkstra, findShortestPath } from "../Algorithms/dijkstra";
import { aStar } from "../Algorithms/a*";

import "./Grid.css";
import "./Node/Node.css";

export default class Grid extends Component {
  state = {
    grid: [],
    fenceToggle: false,
    mouseToggle: false,
    start: {
      present: false,
      gridId: {
        rowIndex: null,
        colIndex: null,
      },
    },
    finish: {
      present: false,
      gridId: {
        rowIndex: null,
        colIndex: null,
      },
    },
    fence: {
      present: [false],
      gridId: {
        rowIndex: null,
        colIndex: null,
      },
    },
  };

  componentDidMount() {
    this.gridSetup();
  }

  createNode = (gridId) => {
    return {
      gridId,
      start: false,
      finish: false,
      distance: Infinity,
      visited: false,
      pastNode: null,
      fence: false,
    };
  };

  gridSetup = () => {
    const { width, height } = this.props.view;

    let grid = [];
    for (let rowIndex = 0; rowIndex < width; rowIndex++) {
      let current_row = [];
      for (let colIndex = 0; colIndex < height; colIndex++) {
        const gridId = { colIndex, rowIndex };

        current_row.push(this.createNode(gridId));
      }
      grid.push(current_row);
    }

    this.setState({ grid });
  };

  fenceToggler = () => {
    const { fenceToggle } = this.state;
    if (fenceToggle === false) {
      this.setState({ fenceToggle: true });
    } else {
      this.setState({ fenceToggle: false });
    }
  };

  mouseFlag = () => {
    const { mouseToggle } = this.state;
    if (mouseToggle === false) {
      this.setState({ mouseToggle: true });
    } else {
      this.setState({ mouseToggle: false });
    }
  };

  nodeFlag = (gridId, type) => {
    const { grid } = this.state;
    const { rowIndex, colIndex } = gridId;
    const node = grid[rowIndex][colIndex];
    if (type === "fence") {
      if (node.start === true || node.finish === true) {
        node[type] = false;
      } else {
        node[type] = true;
      }
    } else {
      node[type] = true;
      this.setState({
        [type]: {
          present: true,
          gridId: gridId,
        },
      });
    }
  };

  // commented out as it is currently unnecessary, may be useful later
  // resetStartFinish = () => {
  //   this.setState({
  //     start: { ...this.state.start, present: false },
  //     finish: { ...this.state.finish, present: false },
  //   });
  // };

  runDijkstra = () => {
    const { grid, start, finish } = this.state;
    const startNode = grid[start.gridId.rowIndex][start.gridId.colIndex];
    const finishNode = grid[finish.gridId.rowIndex][finish.gridId.colIndex];
    const resultOfDijkstra = dijkstra(grid, startNode, finishNode);
    const y = findShortestPath(resultOfDijkstra[resultOfDijkstra.length - 1]);

    this.animateAlgorithm(resultOfDijkstra, y);
  };

  animateAlgorithm = (visitedNodesInOrder, nodesInShortestPathOrder) => {
    for (let i = 0; i <= visitedNodesInOrder.length; i++) {
      if (i === visitedNodesInOrder.length) {
        setTimeout(() => {
          this.animateShortestPath(nodesInShortestPathOrder);
        }, 10 * i);
        return;
      }
      setTimeout(() => {
        const node = visitedNodesInOrder[i];
        document.getElementById(
          `node-${node.gridId.colIndex}-${node.gridId.rowIndex}`
        ).className = `Node ${
          node.start
            ? "start"
            : node.finish
            ? "finish"
            : node.visited
            ? "visited"
            : ""
        }`;
      }, 10 * i);
    }
  };

  animateShortestPath = (nodesInShortestPathOrder) => {
    for (let i = 1; i < nodesInShortestPathOrder.length - 1; i++) {
      setTimeout(() => {
        const node = nodesInShortestPathOrder[i];
        document.getElementById(
          `node-${node.gridId.colIndex}-${node.gridId.rowIndex}`
        ).className = "Node path";
      }, 50 * i);
    }
  };

  render() {
    const { grid, start, finish } = this.state;

    const nodes = grid.map((row, colIndex) => {
      return (
        <div className="Column" key={colIndex.toString()}>
          {row.map((node, rowIndex) => (
            <Node
              key={colIndex.toString() + " " + rowIndex.toString()}
              id={`node-${node.gridId.colIndex}-${node.gridId.rowIndex}`}
              gridId={node.gridId}
              gridHasStart={start.present}
              gridHasFinish={finish.present}
              gridHasFenceToggle={this.state.fenceToggle}
              nodeFlag={this.nodeFlag}
              mouseFlag={this.mouseFlag}
              updateNode={this.updateNode}
              mouseToggle={this.state.mouseToggle}
              // reset={this.resetStartFinish}
            />
          ))}
        </div>
      );
    });

    return (
      <Fragment>
        {start.present && finish.present ? (
          <Header run={this.runDijkstra} fenceToggle={this.fenceToggler} />
        ) : (
          <Alert variant="primary">Please Choose A Start & End Node</Alert>
        )}
        <div>
          <Info />
          <div className="Grid" children={nodes} />
        </div>
      </Fragment>
    );
  }
}

Grid.propTypes = {
  view: PropTypes.object.isRequired,
};
