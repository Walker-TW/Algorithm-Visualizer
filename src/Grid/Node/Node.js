import React, { Component } from "react";
import "./Node.css";
import PropTypes from "prop-types";

export default class Node extends Component {
  state = {
    start: false,
    finish: false,
    fence: false,
  };

  clickHandler = () => {
    const {
      gridId,
      gridHasStart,
      gridHasFinish,
      gridHasFenceToggle,
      nodeFlag,
    } = this.props;

    if (gridHasStart && gridHasFinish && !gridHasFenceToggle) {
      document.getElementById("button").className = "btn btn-primary big";
    } else {
      if (!gridHasStart) {
        nodeFlag(gridId, "start");
        this.setState({ start: true });
      } else {
        this.setState({ finish: true });
        nodeFlag(gridId, "finish");
      }
    }
  };

  fenceSelector = (e) => {
    const { gridId, nodeFlag } = this.props;
    const { fence } = this.state;
    if (fence === false) {
      nodeFlag(gridId, "fence");
      this.setState({ fence: true });
    } else {
      nodeFlag(gridId, "fence");
      this.setState({ fence: false });
    }
  };

  mouseUpHandler = () => {
    const { mouseFlag, mouseToggle, gridHasFenceToggle } = this.props;
    if (gridHasFenceToggle === true && mouseToggle === true) {
      mouseFlag();
    }
  };
  mouseEnterHandler = () => {
    const { mouseToggle, gridHasFenceToggle } = this.props;
    if (gridHasFenceToggle === true && mouseToggle === true) {
      this.fenceSelector();
    }
  };
  mouseDownHandler = () => {
    const { mouseToggle, mouseFlag, gridHasFenceToggle } = this.props;
    if (gridHasFenceToggle === true && mouseToggle === false) {
      mouseFlag();
      this.fenceSelector();
    } else {
      this.clickHandler();
    }
  };

  render() {
    return (
      <div
        onMouseDown={this.mouseDownHandler}
        onMouseEnter={this.mouseEnterHandler}
        onMouseUp={this.mouseUpHandler}
        className={`Node ${
          this.state.start
            ? "start"
            : this.state.finish
            ? "finish"
            : this.state.fence
            ? "fence"
            : ""
        }`}
        id={this.props.id}
      />
    );
  }
}

Node.propTypes = {
  gridId: PropTypes.object.isRequired,
  nodeFlag: PropTypes.func.isRequired,
  gridHasStart: PropTypes.bool.isRequired,
  gridHasFinish: PropTypes.bool.isRequired,
  gridHasFenceToggle: PropTypes.bool.isRequired,

  // reset: PropTypes.func.isRequired,
};
