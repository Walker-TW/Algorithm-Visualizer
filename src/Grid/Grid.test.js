import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount, render } from 'enzyme';

import Grid from './Grid';
import Node from './Node/Node';

describe('<Grid />', () => {
  it('renders', () => {
    const wrapper = render(<Grid />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders all 1500 Node components', () => {
    const wrapper = mount(<Grid />);
    expect(wrapper.find(Node).length).toEqual(100);
  });

  it('render a node that will change the start state when clicked', () => {
    const wrapper = mount(<Grid />);
    const node = wrapper.find(Node);
    const test = node.first();
    expect(test.state().isStart).toEqual(false);
    test.simulate('click');
    expect(test.state().isStart).toEqual(true);
  });

  it('render a node that will change the finish state when clicked', () => {
    const wrapper = mount(<Grid />);
    const node = wrapper.find(Node);
    const test = node.first();
    expect(test.state().isFinish).toEqual(false);
    test.simulate('click');
    test.simulate('click');
    expect(test.state().isFinish).toEqual(true);
  });
});
