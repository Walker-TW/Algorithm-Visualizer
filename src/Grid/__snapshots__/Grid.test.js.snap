// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid view={desktop} /> renders 1`] = `
Array [
  <nav
    class="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <a
      class="navbar-brand"
      href="#home"
    >
      Algo-Visualiser
    </a>
    <button
      aria-controls="responsive-navbar-nav"
      aria-label="Toggle navigation"
      class="navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="navbar-collapse collapse"
      id="responsive-navbar-nav"
    >
      <div
        class="mr-auto navbar-nav"
      >
        <a
          class="nav-link"
          data-rb-event-key="#features"
          href="#features"
        >
          Features
        </a>
        <a
          class="nav-link"
          data-rb-event-key="#pricing"
          href="#pricing"
        >
          Pricing
        </a>
        <div
          class="dropdown nav-item"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle nav-link"
            href="#"
            id="collasible-nav-dropdown"
            role="button"
          >
            Algorithms
          </a>
        </div>
        <form
          class=""
        >
          <div
            class="form-check"
            style="color:white"
          >
            <input
              class="form-check-input"
              id="fence-check"
              name="fences"
              type="checkbox"
            />
            <label
              class="form-check-label"
              for="fence-check"
              title=""
            >
              Fences
            </label>
          </div>
        </form>
      </div>
      <div
        class="navbar-nav nav-center"
      >
        <button
          class="btn btn-primary"
          id="button"
          type="button"
        >
          Let's Run Dijkstra
        </button>
        <button
          class="btn btn-primary"
          id="reset-btn"
          style="background-color:red"
          type="button"
        >
          Reset
        </button>
      </div>
      <div
        class="navbar-nav"
      >
        <a
          class="nav-link"
          data-rb-event-key="#deets"
          href="#deets"
        >
          More deets
        </a>
        <a
          class="nav-link"
          data-rb-event-key="2"
          href="#memes"
        >
          Contact
        </a>
      </div>
    </div>
  </nav>,
  <div
    class="fade alert alert-primary show"
    role="alert"
  >
    Please Choose A Start & End Node
  </div>,
  <div>
    <div
      class="collapsed"
    >
      info
    </div>
    <div
      class="Grid"
    />
  </div>,
]
`;
